src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -371,66 +372,121 @@ private static bool IsMultidimensionalInitializer(ImmutableArray<BoundExpression
+        /// An output Boolean indicating whether a caller trying to perform in-place initialization should instead prefer to assign the local to a new value.
+        /// Call sites may try to optimize an assignment to a newly-created struct by calling the constructor directly rather than assigning, but that
+        /// may then inhibit the more valuable optimization of creating a span via RuntimeHelpers.CreateSpan, which needs to assign. When a caller has passed
+        /// in an <paramref name="inPlaceTarget"/> but CreateSpan could be used if it weren't, this method may return false and set <paramref name="avoidInPlace"/>
+        /// to true to inform the caller it can try again without the <paramref name="inPlaceTarget"/>.
+        /// </param>
+        /// <param name="start">The expression for the offset into the array being wrapped in a span.</param>
+        /// <param name="length">The expression for the length of the subarray being wrapped in a span.</param>
+        /// <returns>
+        /// true if this method successfully emit a ReadOnlySpan as a wrapper for a blob; otherwise, false.  If false, nothing will have been emitted.
+        /// And if false and <paramref name="avoidInPlace"/> is true (in which case <paramref name="inPlaceTarget"/> must have been non-null), the caller
+        /// may try again but with a null <paramref name="inPlaceTarget"/>.
+        /// </returns>
+        private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExpression wrappedExpression, bool used, BoundExpression inPlaceTarget, out bool avoidInPlace, BoundExpression? start = null, BoundExpression? length = null)
         {
-            Debug.Assert(start is null == length is null);
+            Debug.Assert(inPlaceTarget is null || TargetIsNotOnHeap(inPlaceTarget), "in-place construction target should not be on heap");
+            Debug.Assert(_diagnostics.DiagnosticBag is not null, $"Expected non-null {nameof(_diagnostics)}.{nameof(_diagnostics.DiagnosticBag)}");
+
+            if (start is null != length is null)
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
 
-            ImmutableArray<byte> data = default;
             int elementCount = -1;
+            avoidInPlace = false;
+            SpecialType specialElementType = SpecialType.None;
 
             if (!_module.SupportsPrivateImplClass)
             {
                 return false;
             }
 
-            var ctor = (MethodSymbol?)Binder.GetWellKnownTypeMember(this._module.Compilation, WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer, _diagnostics, syntax: wrappedExpression.Syntax, isOptional: true);
-
-            if (ctor == null)
+            var rosPointerCtor = (MethodSymbol?)Binder.GetWellKnownTypeMember(_module.Compilation, WellKnownMember.System_ReadOnlySpan_T__ctor_Pointer, _diagnostics, syntax: wrappedExpression.Syntax, isOptional: true);
+            if (rosPointerCtor is null)
```

Comment:
>rosPointerCtor [](http://example.com/codeflow?start=16&length=14)

Is this helper always needed? Can its absence prevent optimization that is still possible to do?

Reply:
We could fall back to using the array caching scheme if the `ReadOnlySpan<T>(T[])` constructor exists.  But I don't think it's worth complicating the compiler to try to optimize the case of one of the constructors existing but not the other.  Both constructors were both added to `ReadOnlySpan<T>` when it was initially released, in both .NET Core and in the System.Memory nuget package available for use downlevel.  We'd be optimizing for a case that's not meaningful and that would basically never be hit outside of the compiler's tests.